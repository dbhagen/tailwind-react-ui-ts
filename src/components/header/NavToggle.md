Button that toggles the visibility of the `NavMenu` when screen size is small enough to trigger responsive view.
