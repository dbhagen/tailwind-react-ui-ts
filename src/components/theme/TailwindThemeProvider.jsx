import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

const TailwindThemeProvider = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
