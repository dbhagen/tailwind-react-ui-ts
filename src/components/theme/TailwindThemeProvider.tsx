import React from 'react'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

export type TailwindThemeProviderProps = {
  theme: {}
  children: React.ReactNode
}

const TailwindThemeProvider: React.FunctionComponent<TailwindThemeProviderProps> = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

export default TailwindThemeProvider
