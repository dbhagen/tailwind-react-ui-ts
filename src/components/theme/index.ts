export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
