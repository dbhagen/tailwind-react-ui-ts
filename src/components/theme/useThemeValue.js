import { tailwindPropToClassName } from '../tailwind'

export default (prefix, value, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
