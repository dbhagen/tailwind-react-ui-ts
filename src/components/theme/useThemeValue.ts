import { tailwindPropToClassName } from '../tailwind'

export default (prefix: string, value: string, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
