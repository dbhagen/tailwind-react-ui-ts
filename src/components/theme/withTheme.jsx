import React from 'react'

import TailwindTheme from './TailwindTheme'

const withTheme = Component => {
  const WithTheme = props => (
    <TailwindTheme.Consumer>
      {theme => <Component {...props} theme={theme} />}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}

export default withTheme
