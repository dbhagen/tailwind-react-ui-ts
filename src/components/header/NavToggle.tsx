import React from 'react'
import PropTypes from 'prop-types'

import { Box } from '../primitives'
import { Button } from '../button'
import { withTheme } from '../theme'

const Bar = () => (
  <Box inlineBlock border="b" style={{ borderColor: 'currentColor' }} />
)

const NavToggle = ({
  theme,
  children,
  onClick,
  header: { onToggle, style, id, screen },
  ...rest
}) => {
  const handleClick = e => {
    onToggle()
    if (onClick) onClick(e)
  }

  const responsive = screen
    ? {
        [`hidden-${screen}`]: true,
      }
    : {}

  return (
    <Button
      w={12}
      h={12}
      p={0}
      block
      onClick={handleClick}
      aria-label="Open menu"
      aria-haspopup="true"
      aria-controls={`${id}-nav`}
      text={style.text || theme.textColors.on.primary}
      bg-hocus={style.text || theme.textColors.on.primary}
      text-hocus={style.bg || theme.brandColors.primary}
      {...responsive}
      {...rest}
    >
      {children || (
        <Box
          flex={[true, 'col']}
          items="stretch"
          justify="around"
          h="full"
          p={3}
        >
          <Bar />
          <Bar />
          <Bar />
        </Box>
      )}
    </Button>
  )
}

NavToggle.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
  header: PropTypes.shape({
    onToggle: PropTypes.func.isRequired,
    screen: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    style: PropTypes.object,
  }),
  onClick: PropTypes.func,
}

NavToggle.defaultProps = {
  children: undefined,
  onClick: undefined,
  header: {
    style: {},
    screen: 'lg',
  },
}

export { NavToggle as component }
export default withTheme(NavToggle)
