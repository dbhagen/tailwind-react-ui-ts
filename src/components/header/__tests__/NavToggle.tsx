import React from 'react'
import { shallow } from 'enzyme'

import { component as NavToggle } from '../NavToggle'
import { defaultTheme } from '../../theme'

const setup = (testProps = {}) => {
  const props = Object.assign(
    {
      header: { onToggle: jest.fn(), style: { text: 'white' } },
      theme: defaultTheme,
    },
    testProps,
  )

  const wrapper = shallow(<NavToggle {...props} />)

  return {
    props,
    wrapper,
  }
}

describe('NavToggle', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
