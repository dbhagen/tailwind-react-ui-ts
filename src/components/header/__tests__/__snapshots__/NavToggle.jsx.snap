// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavToggle renders matching snapshot 1`] = `
<WithTheme(Button)
  aria-controls="undefined-nav"
  aria-haspopup="true"
  aria-label="Open menu"
  bg-hocus="white"
  block={true}
  h={12}
  onClick={[Function]}
  p={0}
  text="white"
  text-hocus="blue-400"
  w={12}
>
  <Box
    flex={
      Array [
        true,
        "col",
      ]
    }
    h="full"
    inline={false}
    inlineBlock={false}
    is="div"
    items="stretch"
    justify="around"
    p={3}
  >
    <Bar />
    <Bar />
    <Bar />
  </Box>
</WithTheme(Button)>
`;
